process/process.go

```diff
@@ -0,0 +1,380 @@
+	switch m := m.(type) {
+	case *Propose:
+		p.handlePropose(m)
+	case *Prevote:
+		p.handlePrevote(m)
+	case *Precommit:
+		p.handlePrecommit(m)
+	}
+}
+
+func (p *Process) startRound(round block.Round) {
+	p.state.CurrentRound = round
+	p.state.CurrentStep = StepPropose
+	if p.signatory.Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+		var proposal block.Block
+		if p.state.ValidBlock.Hash() != block.InvalidHash {
+			proposal = p.state.ValidBlock
+		} else {
+			proposal = p.proposer.BlockProposal(p.state.CurrentHeight, p.state.CurrentRound)
+		}
+		p.broadcaster.Broadcast(NewPropose(
+			p.state.CurrentHeight,
+			p.state.CurrentRound,
+			proposal,
+			p.state.ValidRound,
+		))
+	} else {
+		p.scheduleTimeoutPropose(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPropose, p.state.CurrentRound))
+	}
+}
+
+func (p *Process) handlePropose(propose *Propose) {
+	_, firstTime, _, _ := p.state.Proposals.Insert(propose)
+
+	// upon Propose{currentHeight, currentRound, block, -1}
+	if propose.Height() == p.state.CurrentHeight && propose.Round() == p.state.CurrentRound && propose.ValidRound() == block.InvalidRound {
+		// from Schedule{currentHeight, currentRound}
+		if propose.Signatory().Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+			// while step = StepPropose
+			if p.state.CurrentStep == StepPropose {
+				if p.validator.IsBlockValid(propose.Block()) && (p.state.LockedRound == block.InvalidRound || p.state.LockedBlock.Equal(propose.Block())) {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						propose.Block().Hash(),
+					))
+				} else {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						block.InvalidHash,
+					))
+				}
+				p.state.CurrentStep = StepPrevote
+			}
+		}
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
+	if firstTime {
+		p.checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime()
+	}
+	p.checkProposeInCurrentHeightWithPrecommits(propose.Round())
+}
+
+func (p *Process) handlePrevote(prevote *Prevote) {
+	n, _, _, firstTimeExceeding2F := p.state.Prevotes.Insert(prevote)
+	if firstTimeExceeding2F && prevote.Height() == p.state.CurrentHeight && prevote.Round() == p.state.CurrentRound && p.state.CurrentStep == StepPrevote {
+		// upon 2f+1 Prevote{currentHeight, currentRound, *} while step = StepPrevote for the first time
+		p.scheduleTimeoutPrevote(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPrevote, p.state.CurrentRound))
+	}
+
+	// upon f+1 *{currentHeight, round, *, *} and round > currentRound
+	if n > p.state.Prevotes.F() && prevote.Height() == p.state.CurrentHeight && prevote.Round() > p.state.CurrentRound {
+		p.startRound(prevote.Round())
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
+	if firstTimeExceeding2F {
+		p.checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime()
+	}
```

Comment:
Lines 44 - 46 from the paper seem to be missing.

Reply:
@ross-pure thanks, pushed a fix.